### PR TITLE
8315462: [REDO] runtime/NMT/SummarySanityCheck.java failed with "Total committed (MMMMMM) did not match the summarized committed (NNNNNN)"

### PR DESCRIPTION
**Pull request by afshin-zafari**
When taking snapshot of malloc entries, other threads may update the items that already copied so change their corresponding sum/count. To make the sum/count values be consistent with whatever copied, they are calculated during copy and then set explicitly if any mismatch detected.
The function body is also moved to cpp file.  

### Tests
mach5 tiers 1-5 all platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315462](https://bugs.openjdk.org/browse/JDK-8315462): [REDO] runtime/NMT/SummarySanityCheck.java failed with "Total committed (MMMMMM) did not match the summarized committed (NNNNNN)" (**Bug** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [a1014001](https://git.openjdk.org/jdk/pull/16237/files/a1014001379afdb2c638ece81f4b18b02462fcd9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16237/head:pull/16237` \
`$ git checkout pull/16237`

Update a local copy of the PR: \
`$ git checkout pull/16237` \
`$ git pull https://git.openjdk.org/jdk.git pull/16237/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16237`

View PR using the GUI difftool: \
`$ git pr show -t 16237`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16237.diff">https://git.openjdk.org/jdk/pull/16237.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16237#issuecomment-1767988387)

### CODE REVIEW

File: src/hotspot/share/services/mallocTracker.cpp [hunk 1]
```diff
@@ -59,6 +59,26 @@ void MemoryCounter::update_peak(size_t size, size_t cnt) {
...
+  s->_all_mallocs = _all_mallocs;
+  size_t total_size = 0;
+  size_t total_count = 0;
+  for (int index = 0; index < mt_number_of_types; index ++) {
+    s->_malloc[index] = _malloc[index];
+    total_size += s->_malloc[index].malloc_size();
+    total_count += s->_malloc[index].malloc_count();
+  }
+  // ThreadCritical is used only for Virtual Memory de-/allocations.
+  // So, it is possible that concurrent malloc/free happened during the copy to snapshot.
```
Comment (tstuefe) [inline src/hotspot/share/services/mallocTracker.cpp:76]: I'd shorten this to something like "malloc counters may be updated concurrently", that's succinct enough. Also, we may switch to mutexes at some point, like we should, then this commend would be outdated.

File: src/hotspot/share/services/mallocTracker.cpp [hunk 2]
```diff
@@ -59,6 +59,26 @@ void MemoryCounter::update_peak(size_t size, size_t cnt) {
...
+  if (s->_all_mallocs.size() != total_size || s->_all_mallocs.count() != total_count) {
+    s->_all_mallocs.set_size_and_count(total_size, total_count);
+  }
```
Comment (tstuefe) [inline src/hotspot/share/services/mallocTracker.cpp:79]: Why just not do it unconditionally?

**Review by tstuefe — APPROVED** [review]
Good. Minor nits. Up to you if you change anything.

Cheers, Thomas

File: src/hotspot/share/services/mallocTracker.cpp (hunk 2 above)
Comment (afshin-zafari) [reply src/hotspot/share/services/mallocTracker.cpp]: Yes. Changed.

**Comment by afshin-zafari** [comment]
A new [PR](https://github.com/openjdk/jdk/pull/16262) is created to continue.

**Comment by jdksjolen** [comment]
If the new PR supersedes the old one, then can this be closed?

**Comment by afshin-zafari** [comment]
> If the new PR supersedes the old one, then can this be closed?

Yes. It can. I close it.

**Comment by afshin-zafari** [comment]
A new [PR](https://github.com/openjdk/jdk/pull/16262) is created to continue.
